lib/posix/fs.c

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <string.h>
+#include <errno.h>
+#include <posix/unistd.h>
+
+
+struct p_fs_file {
+	struct fs_file_t file;
+	bool used;
+};
+
+struct p_fs_dir {
+	struct fs_dir_t	dir;
+	bool used;
+};
+
+static struct p_fs_dir dir_array[CONFIG_MAX_OPEN_FILES];
+K_MUTEX_DEFINE(dirp_mutex);
+
+static struct p_fs_file file_array[CONFIG_MAX_OPEN_FILES];
+K_MUTEX_DEFINE(filep_mutex);
+
+static struct fs_dirent fdirent;
+static struct dirent pdirent;
+
+/**
+ * @brief Open a file.
+ *
+ * See IEEE 1003.1
+ */
+int open(const char *name, int flags)
+{
+	int rc, fd;
+	struct fs_file_t *ptr;
+
+	ARG_UNUSED(flags);
+
+	k_mutex_lock(&filep_mutex, K_FOREVER);
+	/* Find free file descriptor */
+	for (fd = 0; fd < CONFIG_MAX_OPEN_FILES; fd++) {
+		if (file_array[fd].used == false) {
+			ptr = &file_array[fd].file;
+			memset(ptr, 0, sizeof(struct fs_file_t));
+			file_array[fd].used = true;
+			break;
+		}
+	}
+	k_mutex_unlock(&filep_mutex);
+
+	if (fd >= CONFIG_MAX_OPEN_FILES) {
+		return -1;
+	}
+
+	/* Allocate memory for file ptr */
+
+	rc = fs_open(ptr, name);
+	if (rc < 0) {
+		k_mutex_lock(&filep_mutex, K_FOREVER);
+		file_array[fd].used = false;
+		k_mutex_unlock(&filep_mutex);
+		return -1;
+	}
+
+	return fd;
+}
+
+/**
+ * @brief Close a file descriptor.
+ *
+ * See IEEE 1003.1
+ */
+int close(int fd)
+{
+	int rc;
+
+	if (fd >= CONFIG_MAX_OPEN_FILES) {
+		return -EBADF;
+	}
+
+	rc = fs_close(&file_array[fd].file);
+
+	/* Free file ptr memory */
+	k_mutex_lock(&filep_mutex, K_FOREVER);
+	file_array[fd].used = false;
+	k_mutex_unlock(&filep_mutex);
+
+	return rc < 0 ? -1 : 0;
+}
+
+/**
+ * @brief Write to a file.
+ *
+ * See IEEE 1003.1
+ */
+ssize_t write(int fd, char *buffer, unsigned int count)
+{
+	int rc;
+
+	rc = fs_write(&file_array[fd].file, buffer, count);
+	return rc < 0 ? -1 : rc;
+}
+
+/**
+ * @brief Read from a file.
+ *
+ * See IEEE 1003.1
+ */
+ssize_t read(int fd, char *buffer, unsigned int count)
+{
+	int rc;
+
+	rc = fs_read(&file_array[fd].file, buffer, count);
+	return rc < 0 ? -1 : rc;
+}
+
+/**
+ * @brief Move read/write file offset.
+ *
+ * See IEEE 1003.1
+ */
+int lseek(int fd, int offset, int whence)
+{
+	int rc;
+
+	rc = fs_seek(&file_array[fd].file, offset, whence);
+	return rc < 0 ? -1 : rc;
+}
+
+/**
+ * @brief Open a directory stream.
+ *
+ * See IEEE 1003.1
+ */
+DIR *opendir(const char *dirname)
+{
+	int rc, fd;
+	struct fs_dir_t *ptr;
+
+	k_mutex_lock(&dirp_mutex, K_FOREVER);
+	/* Allocate memory for dir ptr */
+	for (fd = 0; fd < CONFIG_MAX_OPEN_FILES; fd++) {
+		if (dir_array[fd].used == false) {
+			ptr = &dir_array[fd].dir;
+			memset(ptr, 0, sizeof(struct fs_dir_t));
+			dir_array[fd].used = true;
+			break;
```

Comment:
This loop is identical to the "find an open slot" code in open() above.  Please unify.

It's also susceptible to the same criticism about locking I had there, which I won't type twice because it should only appear once in the file anyway. :)